File: src/platform/posix/time.c
/* public domain, no copyright claimed */

#include <stdlib.h>
#include <errno.h>
#include <assert.h>
#include <time.h>
#include <math.h>
#include <unistd.h>

#include <stdint.h>
#include <stdbool.h>

#ifdef __OpenBSD__
#include <sys/types.h>
#include <sys/sysctl.h>
#endif

#include "../platform_types.h"

long long int arcan_timemillis()
{
	struct timespec tp;
	clock_gettime(CLOCK_MONOTONIC, &tp);
	return (tp.tv_sec * 1000) + (tp.tv_nsec / 1000000);
}

long long int arcan_timemicros()
{
	struct timespec tp;
	clock_gettime(CLOCK_MONOTONIC, &tp);
	return (tp.tv_sec * 1000000) + (tp.tv_nsec / 1000);
}

struct platform_timing platform_hardware_clockcfg()
{
#ifdef __OpenBSD__
	int mib[2] = {CTL_KERN, KERN_CLOCKRATE};
	struct clockinfo cinf;
	size_t len = sizeof(struct clockinfo);
	if (sysctl(mib, 2, &cinf, &len, NULL, 0) != -1){
		return (struct platform_timing){
			.cost_us = cinf.tick,
			.tickless = false
		};
	}
	else {
		return (struct platform_timing){
			.cost_us = 10 * 1000,
			.tickless = false
		};
	}
#else
	return (struct platform_timing){
		.cost_us = 0,
		.tickless = true
	};
#endif
}

void arcan_timesleep(unsigned long val)
{
	struct timespec req, rem;
	req.tv_sec = floor(val / 1000);
	val -= req.tv_sec * 1000;
	req.tv_nsec = val * 1000000;

	while( nanosleep(&req, &rem) == -1 ){
		assert(errno != EINVAL);
		if (errno == EFAULT)
			break;

/* sweeping EINTR introduces an error rate that can grow large,
 * check if the remaining time is less than a threshold */
		if (errno == EINTR) {
			req = rem;
			if (rem.tv_sec * 1000 + (1 + req.tv_nsec) / 1000000 < 4)
				break;
		}
	}
}

File: src/platform/posix/time.zig
// Zig port of posix/time.c
// Monotonic clock queries (ms/us), platform timing config, and EINTR-safe nanosleep.

const std = @import("std");
const builtin = @import("builtin");
const is_freestanding = (builtin.os.tag == .freestanding);

const platform_timing = extern struct {
    tickless: bool,
    cost_us: c_uint,
};

export fn arcan_timemillis() c_longlong {
    if (is_freestanding) return 0;
    return arcan_timemillis_posix();
}

export fn arcan_timemicros() c_longlong {
    if (is_freestanding) return 0;
    return arcan_timemicros_posix();
}

export fn platform_hardware_clockcfg() platform_timing {
    return .{
        .cost_us = 0,
        .tickless = true,
    };
}

export fn arcan_timesleep(val: c_ulong) void {
    if (is_freestanding) return;
    arcan_timesleep_posix(val);
}

// --- POSIX implementations (only compiled on non-freestanding) ---

fn arcan_timemillis_posix() c_longlong {
    var tp: std.c.timespec = undefined;
    _ = std.c.clock_gettime(std.c.CLOCK.MONOTONIC, &tp);
    return @as(c_longlong, tp.sec) * 1000 + @divTrunc(tp.nsec, 1_000_000);
}

fn arcan_timemicros_posix() c_longlong {
    var tp: std.c.timespec = undefined;
    _ = std.c.clock_gettime(std.c.CLOCK.MONOTONIC, &tp);
    return @as(c_longlong, tp.sec) * 1_000_000 + @divTrunc(tp.nsec, 1000);
}

fn arcan_timesleep_posix(val: c_ulong) void {
    var ms = val;
    const secs: isize = @intCast(ms / 1000);
    ms -= @as(c_ulong, @intCast(secs)) * 1000;
    var req: std.c.timespec = .{
        .sec = secs,
        .nsec = @intCast(ms * 1_000_000),
    };

    var rem: std.c.timespec = .{ .sec = 0, .nsec = 0 };

    while (std.c.nanosleep(&req, &rem) == -1) {
        const e = std.c._errno().*;
        std.debug.assert(e != @intFromEnum(std.os.linux.E.INVAL));
        if (e == @intFromEnum(std.os.linux.E.FAULT))
            break;

        // Sweeping EINTR introduces an error rate that can grow large;
        // check if the remaining time is less than a threshold (4 ms).
        if (e == @intFromEnum(std.os.linux.E.INTR)) {
            req = rem;
            const rem_ms = rem.sec * 1000 + @divTrunc(1 + req.nsec, 1_000_000);
            if (rem_ms < 4)
                break;
        }
    }
}
